### PR TITLE
Add tests for new tracker blocking report parameters

### PR DESCRIPTION
**Pull request by GuiltyDolphin**
Asana: https://app.asana.com/0/0/1204271046995902/f

This adds the following tests:

- check for remote config parameters (all platforms)
- check for description encoding (platforms that currently support descriptions)
- check for correct encoding of new tracker blocking parameters (extension)
- check for truncation of request parameters on long reports (extension)

### CODE REVIEW

File: broken-site-reporting/README.md [hunk 1]
```diff
@@ -14,6 +14,10 @@ Test suite specific fields:
 - `wasUpgraded` - bool - if request was upgraded to HTTPS by us or not
 - `category` - string - user picked breakage category e.g. 'content', 'images', 'paywall', 'login'
 - `blockedTrackers` - array of strings - array of hostnames of trackers that were blocked
+- `noActionRequests` - array of strings - array of hostnames of requests that had `none` action
+- `adAttributionRequests` - array of strings - array of hostnames of requests that had `ad-attribution` action
+- `ignoredByUserRequests` - array of strings - array of hostnames of requests that had `ignore-user` action
```
Comment (kdzwinel) [inline broken-site-reporting/README.md:19]: none/ad-attribution/ignore-user - those are extension specific names, I don't think we should use them here, but rather explain what they mean

File: broken-site-reporting/tests.json [hunk 2]
```diff
@@ -74,6 +104,102 @@
...
+                "exceptPlatforms": [
+                    "android-browser",
+                    "ios-browser",
+                    "macos-browser",
+                    "safari-extension",
+                    "windows-browser"
+                ]
+            },
+            {
+                "name": "Test truncation when report is too long (URL)",
```
Comment (kdzwinel) [inline broken-site-reporting/tests.json:143]: nice, is truncation documented? only extension supports it ATM?

**Review by kdzwinel — COMMENTED** [review]
Thanks a lot Ben, appreciate it. I did a first pass. Those are passing against the extension?

File: broken-site-reporting/tests.json (hunk 2 above)
Comment (GuiltyDolphin) [reply broken-site-reporting/tests.json:143]: @kdzwinel I'm adding truncation documentation to the guidelines. It's only extension at the moment

File: broken-site-reporting/tests.json [hunk 3]
```diff
@@ -74,6 +104,102 @@
...
+                "expectReportURLParams": [
+                    {"name": "category", "value": "content"},
+                    {"name": "siteUrl", "value": "https%3A%2F%2Fexample.test%2F"},
+                    {"name": "upgradedHttps", "value": "true"},
+                    {"name": "tds", "value": "abc123"},
+                    {"name": "blockedTrackers", "value": "bad.tracker.test,tracking.test"},
+                    {"name": "ignoreRequests", "value": "ignored1.test,sub.ignored2.test"},
+                    {"name": "noActionRequests", "value": "noaction1.test,sub.noaction2.test"},
+                    {"name": "adAttributionRequests", "value": "adattr1.test,sub.adattr2.test"},
+                    {"name": "ignoredByUserRequests", "value": "ignoreuser1.test,sub.ignoreuser2.test"},
```
[the remark this answers is not in the document]
Comment (GuiltyDolphin) [reply broken-site-reporting/tests.json:131]: @kdzwinel for the most part. Due to the way the resolution works, you could end up with e.g., both `ignoredByUserRequests` and `ignoreRequests` populated (blocklist checked prior to user ignore).

I could add another test with empty fields then populated `ignoredByUserRequests` if you think it's helpful?

File: broken-site-reporting/README.md [hunk 4]
```diff
@@ -23,6 +27,9 @@ Test suite specific fields:
 - `gpcEnabled` - boolean - if GPC is enabled or not (native apps only) - GPC can be disabled by user or by remote config
 - `expectReportURLPrefix` - string - resulting report URL should be prefixed with this string
 - `expectReportURLParams` - Array of `{name: '', value: ''}` objects - resulting report URL should have the following set of URL parameters with matching values
+- `remoteConfigEtag` - string - string representation of remote configuration etag
+- `remoteConfigVersion` - string - string representation of remote configuration version
```
[the remark this answers is not in the document]
Comment (GuiltyDolphin) [reply broken-site-reporting/README.md:31]: oh that's a good point - yeah I can see `v1` vs the numeric version being confusing (it's the latter). I'll clarify that here and in the docs

**Comment by GuiltyDolphin** [comment]
@kdzwinel 

> passing against extension

The extension tests need to be updated (cc @jonathanKingston). The long-URL truncation test is passing, but the rest are failing as the parameters aren't currently threading through (see [test file](https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/7097a0103ccfb9410b8c075b08edc77e9b4c46dd/unit-test/background/reference-tests/broken-site-reporting-tests.js)).

File: broken-site-reporting/tests.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (GuiltyDolphin) [reply broken-site-reporting/tests.json:131]: Added a test.

File: broken-site-reporting/README.md [hunk 5]
```diff
@@ -53,4 +60,15 @@ for $testSet in test.json
...
+## Platform exceptions
+
+Please see https://app.asana.com/0/1198207348643509/1200191185434196/f for
+current platform support and further links. A summary of exceptions affecting
+the reference tests is provided below:
+
+- description tests:
+  - the following platforms do not currently allow users to provide a description: `android-browser`, `ios-browser`, `safari-extension`
+- truncation tests:
+  - the following platforms do not currently implement the optional `noActionRequests`, `adAttributionRequests`, `ignoredByUserRequests`, or `ignoreRequests` truncatable parameters: `android-browser`, `ios-browser`, `macos-browser`, `safari-extension`, `windows-browser`
```
Comment (kdzwinel) [inline broken-site-reporting/README.md:74]: something seems wrong here - you have both optional params and turntable params in one sentence, I think you meant to split it somehow?

File: broken-site-reporting/README.md [hunk 6]
```diff
@@ -14,6 +14,10 @@ Test suite specific fields:
 - `wasUpgraded` - bool - if request was upgraded to HTTPS by us or not
 - `category` - string - user picked breakage category e.g. 'content', 'images', 'paywall', 'login'
 - `blockedTrackers` - array of strings - array of hostnames of trackers that were blocked
+- `noActionRequests` - array of strings - array of hostnames of requests that were not on the blocklist
+- `adAttributionRequests` - array of strings - array of hostnames of requests that were ignored due to ad-attribution
+- `ignoredByUserRequests` - array of strings - array of hostnames of requests that were allowed due to protections disabled by the user
+- `ignoreRequests` - array of strings - array of hostnames of requests that were ignored for other reasons (e.g., blocklist ignore)
```
Comment (kdzwinel) [inline broken-site-reporting/README.md:20]:
```suggestion
- `ignoreRequests` - array of strings - array of hostnames of requests that were ignored for other reasons (e.g. blocklist ignore)
```

File: broken-site-reporting/README.md (hunk 5 above)
Comment (GuiltyDolphin) [reply broken-site-reporting/README.md:74]: @kdzwinel see the documentation - all the truncatable parameters are currently optional. The requirement is that if a platform were to implement one (or more) of those optional parameters, then they must also implement truncation.

**Comment by GuiltyDolphin** [comment]
> Thanks Ben, I left two more comments. Do we have a branch of the extension where those tests are passing? Just to make sure that there is no some kind of typo in the tests that will cause them to fail and require another PR to fix.

@kdzwinel not as it stands. The `Test truncation when report is too long (URL)` test is passing, but the rest need some plumbing.

I can open a PR with the plumbing for description and possibly config version, but I'm currently anticipating that @jonathanKingston will update the reference test integration for the remaining parameters as author of the original PR.

File: broken-site-reporting/README.md [hunk 7]
```diff
@@ -53,4 +60,15 @@ for $testSet in test.json
     if $test.expectReportURLParams
         for $param in $test.expectReportURLParams
             expect($url.matchesRegex(/[?&] + $param.name + '=' + $param.value + [&$]/))
-```
\ No newline at end of file
+```
+
+## Platform exceptions
+
+Please see https://app.asana.com/0/1198207348643509/1200191185434196/f for
```
[the remark this answers is not in the document]
Comment (GuiltyDolphin) [reply broken-site-reporting/README.md:67]: @kdzwinel

> make sure we have a task for each of those exceptions

Which exceptions are you referring to? The description exceptions are already linked in that task. Truncation has it's own task in Asana which explains when you should/shouldn't implement it.

> link individual tests

How much are team members using the links in the reference tests vs. the documentation in Asana? (I don't see any benefit to external users as the links are private)

The current flow would be:

1. Visit that Asana link
2. That link is to a task which describes current platform support, with appropriate links elsewhere in Asana

If Asana is _not_ used as the primary source of truth, then we could just document the platform support publicly and point to that documentation from Asana instead.

If Asana _is_ the primary source of truth, then we can try and reduce the chances of this public documentation becoming misaligned by reducing the duplicated information and links.

If we just want to ensure that every exception has a documented issue or Asana task, is it worth documenting those exceptions in the tests themselves (i.e., a link to a GitHub issue or Asana task, like we do for the privacy configuration)? Then we'd have a supporting reason per test exception, and an easy way to find it from a given test.

**Comment by GuiltyDolphin** [comment]
PR with passing tests now merged -> https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/1855

**Review by kdzwinel — APPROVED** [review]
Thank you for addressing all my asks! Nicely done.

File: broken-site-reporting/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kdzwinel) [reply broken-site-reporting/README.md:67]: Per our discussion:
- we want a strong connection between test and task so it's easy to figure out if/when/which exception can be removed and where exception is tracked
- ideally we would keep those links in the JSON next to the exception, but changing tests schema will be painful - we can consider adding a new field
- linking to one asana task where we have all bugs doesn't make it clear which exception is connected to which task + it's not a standard just yet, we can also consider it
